File: test_cases/c/reference/05.c
#include <stdlib.h>
#include <string.h>

int longestValidParentheses(char* s) {
    int n = strlen(s);
    int max = 0;
    int* st = malloc(sizeof(int) * (n + 1));
    int top = 0;
    st[top++] = -1;
    for (int i = 0; i < n; i++) {
        if (s[i] == '(') {
            st[top++] = i;
        } else {
            top--;
            if (top == 0) {
                st[top++] = i;
            } else {
                int len = i - st[top - 1];
                if (len > max) max = len;
            }
        }
    }
    free(st);
    return max;
}

File: test_cases/python/reference/05.py
def longestValidParentheses(s):
    stack = [-1]
    best = 0
    for i, ch in enumerate(s):
        if ch == '(':
            stack.append(i)
        else:
            stack.pop()
            if stack:
                best = max(best, i - stack[-1])
            else:
                stack.append(i)
    return best
